### PR TITLE
Extend -Cdebuginfo with new options and named aliases

### LINKED ISSUE #60020: Rust 1.34 generates significantly less debug information for libstd functions vs. Rust 1.33
A Firefox developer filed [a bug report](https://bugzilla.mozilla.org/show_bug.cgi?id=1544416) about Rust symbols not being correctly represented in crash reports:

```
0 	XUL 	GeckoCrash 	toolkit/xre/nsAppRunner.cpp:5093 	context
1 	XUL 	gkrust_shared::panic_hook 	toolkit/library/rust/shared/lib.rs:240 	frame_pointer
2 	XUL 	core::ops::function::Fn::call 	src/libcore/ops/function.rs:69 	cfi
3 	XUL 	rust_panic_with_hook 	src/libstd/panicking.rs:482 	cfi
4 	XUL 	continue_panic_fmt 	src/libstd/panicking.rs:385 	cfi
5 	XUL 	rust_begin_panic 	src/libstd/panicking.rs:312 	cfi
6 	XUL 	panic_fmt 	src/libcore/panicking.rs:85 	cfi
7 	XUL 	panic 	src/libcore/panicking.rs:49 	cfi
```

The expectation was that, e.g. `rust_panic_with_hook` would have been `std::panicking::rust_panic_with_hook`.

We had recently upgraded to Rust 1.34, which led to comparing object files before and after the upgrade.  I compared Linux x86-64 binaries; I don't see why the analysis doesn't apply to OS X's Mach-O files, but it's possible the results are different there.  (The above crash is from OS X, and we have crashes using Rust 1.33 that do display `std::panicking::rust_panic_with_hook` and similar.)  The ELF symbol table in both cases lists `rust_panic_with_hook` as `_ZN3std9panicking20rust_panic_with_hook$UNIQUE_ID`, so that wasn't the problem.

We then looked at the debug information.  Rust 1.33 generated, according to `readelf --debug-dump=info`:

```
 <3><2155a724>: Abbrev Number: 301 (DW_TAG_subprogram)
    <2155a726>   DW_AT_low_pc      : 0x5738510
    <2155a72a>   DW_AT_high_pc     : 0x6b2
    <2155a72e>   DW_AT_frame_base  : 1 byte block: 54 	(DW_OP_reg4 (esp))
    <2155a730>   DW_AT_linkage_name: (indirect string, offset: 0x9d7fc09): _ZN3std9panicking20rust_panic_with_hook17he447c38467745511E
    <2155a734>   DW_AT_name        : (indirect string, offset: 0x9d7fc45): rust_panic_with_hook
    <2155a738>   DW_AT_decl_file   : 11
    <2155a739>   DW_AT_decl_line   : 447
    <2155a73b>   DW_AT_external    : 1
    <2155a73b>   DW_AT_noreturn    : 1
```

Notice the existence of both `DW_AT_name` and `DW_AT_linkage_name`.  Rust 1.34, in contrast, generated:

```
 <1><216d7afa>: Abbrev Number: 293 (DW_TAG_subprogram)
    <216d7afc>   DW_AT_low_pc      : 0x5734990
    <216d7b00>   DW_AT_high_pc     : 0x6ae
    <216d7b04>   DW_AT_name        : (indirect string, offset: 0x9db1282): rust_panic_with_hook
```

which drops the `DW_AT_linkage_name` and is also significantly less informative than its predecessor.

I'm not familiar enough with `rustc` to know what might have caused this regression.  One of my colleagues pointed out #58208, which changed how various bits of panic infrastructure are imported into `libstd`.  It's not clear to me whether it's that specific change, or how the compiler internally describes `crate::` symbols to LLVM, or something else entirely.

cc @glandium @michaelwoerister

### LINKED ISSUE #64405: debuginfo: Line-tables-only mode emits much more debuginfo than Clang
We probably could emit less debuginfo when compiling with `-Cdebuginfo=1` than we do at the moment. Clang 8.0 emits a lot less debuginfo when compiling with `-gline-tables-only`, with debuginfo not containing any `DW_TAG_subprogram` entries; while stepping through the code with gdb still works.

However, reducing the amount of debuginfo emitted should be done with care since external tools rely on the current behavior (see e.g. #60020).

### PR DESCRIPTION
**Pull request by jyn514**
This is a rebase of https://github.com/rust-lang/rust/pull/83947, along with my best guess at what the new options mean. I tried to follow the LLVM source code to get a better idea but ran into quite a lot of trouble (https://rust-lang.zulipchat.com/#narrow/stream/187780-t-compiler.2Fwg-llvm/topic/go-to-definition.20in.20src.2Fllvm-project.3F). The description for the original PR follows below.

Note that the changes in this PR have already been through FCP: https://github.com/rust-lang/rust/pull/83947#issuecomment-878384979

Closes https://github.com/rust-lang/rust/pull/109311. Helps with https://github.com/rust-lang/rust/pull/104968.
r? @michaelwoerister cc @cuviper

---

The -Cdebuginfo=1 option was never line tables only and can't be due to backwards compatibility issues. This was clarified and an option for emitting line tables only was added. Additionally an option for emitting line info directives only was added, which is needed for some targets, i.e. nvptx. The debug info options should now behave similarly to clang's debug info options.

Fix https://github.com/rust-lang/rust/issues/60020
Fix https://github.com/rust-lang/rust/issues/64405

### CODE REVIEW

File: src/doc/rustc/src/codegen-options/index.md [hunk 1]
```diff
@@ -71,9 +71,11 @@ If not specified, debug assertions are automatically enabled only if the
 This flag controls the generation of debug information. It takes one of the
 following values:
 
-* `0`: no debug info at all (the default).
-* `1`: line tables only.
-* `2`: full debug info.
+* `0` or `none`: no debug info at all (the default).
+* `line-directives-only`: line info directives only. For the nvptx* targets this enables [profiling](https://reviews.llvm.org/D46061). On other targets this behaves the same as `none`.
```
Comment (jyn514) [inline src/doc/rustc/src/codegen-options/index.md:75]: @jdtatz do you know if this second sentence here is correct?

**Comment by michaelwoerister** [comment]
Thanks, @jyn514! I'll take a closer look early next week.

**Comment by cuviper** [comment]
@rust-lang/cargo may also want to think about exposing this in `Cargo.toml`.

File: src/doc/rustc/src/codegen-options/index.md (hunk 1 above)
Comment (michaelwoerister) [reply src/doc/rustc/src/codegen-options/index.md:75]: This does not seem to be true. Looking at the output of `objdump --section-header` and `objdump -W`, there are a few differences between `none`, `line-directives-only`, and `line-tables-only`:
- `none` is the smallest and contains no debuginfo related sections
- `line-directives-only` contains the `.debug_line` section and (an empty) `.debug_aranges`.
- `line-tables-only` adds `.debug_abbrev`, `.debug_info`, `.debug_str`, and a proper `.debug_aranges` section.

So `line-directives-only` is a reduced (degraded?) version of `line-tables-only`. I could imagine that for most tools `line-directives-only` is effectively the same as `none`.

In terms of documentation, my suggestion would be 
```
* `line-directives-only`: line info directives only; a stripped down version of `line-tables-only`. For the nvptx* targets this enables [profiling](https://reviews.llvm.org/D46061). For other use cases, `line-tables-only` is the better, more compatible choice.
```

**Comment by michaelwoerister** [comment]
@bors r+

Thanks for pushing this over the finish line, @jyn514!
And thanks for the original implementation, @jdtatz!

**Pull request merged**